src/libclipper/include/clipper/future.hpp

```diff
@@ -0,0 +1,105 @@
+#ifndef CLIPPER_LIB_FUTURE_HPP
+#define CLIPPER_LIB_FUTURE_HPP
+
+#include <atomic>
+#include <cassert>
+#include <memory>
+#include <thread>
+#include <vector>
+
+#include "boost/thread.hpp"
+
+namespace clipper {
+
+namespace future {
+
+/*
+ * The first element in the pair is a future that will complete when
+ * all the futures provided have completed. The second element is a vector
+ * of futures that have the same values and will complete at the same time
+ * as the futures passed in as argument.
+ *
+ * \param futures A list of futures to compose
+ * \param num_completed A counter indicating the number of futures that
+ * have completed so far. The reason this is an argument to the function
+ * instead of internal state is that the caller of `when_all` must make sure
+ * that the counter stays in scope for the lifetime of all the futures.
+*/
+template <class T>
+std::pair<boost::future<void>, std::vector<boost::future<T>>> when_all(
+    std::vector<boost::shared_future<T>> futures,
+    std::shared_ptr<std::atomic<int>> num_completed) {
+  if (futures.size() == 0) {
+    return std::make_pair(boost::make_ready_future(),
+                          std::vector<boost::future<T>>{});
+  }
+  int num_futures = futures.size();
+  auto completion_promise = std::make_shared<boost::promise<void>>();
+  std::vector<boost::future<T>> wrapped_futures;
+  for (auto f = futures.begin(); f != futures.end(); ++f) {
+    wrapped_futures.push_back(f->then(
```

Comment:
Thanks for the explanation. Your benchmark is definitely a compelling demonstration of the improvement made by using the new `when_all()` and `when_any()` implementations. I found that the time to completion for 100000 requests with the new implementation is roughly half the time for the same number of requests with the boost implementation.